### PR TITLE
First-order meta-gradient every step: 3.183 val loss in 59.3m

### PR DESCRIPTION
**Pull request by dangxingyu**
Each optimizer step splits its own batch in two halves. The gradient of the first half is taken at
the current weights, a temporary step of fixed L2 norm is applied along it on all Muon-managed
matrices, the gradient of the second half is taken at that adapted point, the temporary step is
undone, and the optimizer receives the mean of the two gradients. The second half's gradient at the
adapted point is a first-order meta-gradient: the first half is the adaptation batch, the second half
the evaluation batch. (The script's flags call the episode `ttt`, training-time testing.) Everything is rank-local, there is no extra communication, and the two
half-batches cost exactly what the eight micro-batches already cost; the episode's own work is about
12 ms of a 1.5-1.9 s step. Nothing else in the recipe's data, passes, optimizer, schedules, SWA, or
evaluation changes. Two recipe hyperparameters move with it and are the script's new defaults:
dropout 0.1 -> 0.05 and stochastic depth 0.05 -> 0 (the episode regularises on its own).

Per step (the script's defaults):

```
g   = 1/2 * mean-grad(theta; micro-batches 1-4)       # accumulated in .grad on this rank
d   = g|P / ||g|P||                                    # P = the 279 Muon matrices (1.26B params)
theta_P -= eta * d,  eta = 0.5 * m(t)                  # m(t): the recipe's warmdown LR multiplier
g  += 1/2 * mean-grad(theta; micro-batches 5-8)       # second half at the adapted point
theta_P += eta * d                                     # restore (fp32, to rounding; see below)
optimizer.step(g)                                      # g = 1/2 (grad_A + grad_Q): the usual full-batch mean
```

The restore covers only the matrices this rank owns in the Muon step; the optimizer's all-gather
overwrites the rest, so the restore is exact (two 60-step runs with full and owned restore end at the
same loss to 1e-6) and the extra memory is one eighth of the plastic gradients (0.6 GB).

Results on one 8xH100 80GB HBM3 node, 11 passes, val loss (training time):

| seed | baseline | this script |
| --- | --- | --- |
| 42 | 3.2005 (59.2 min) | 3.1814 (59.2 min) |
| 43 | 3.2013 (59.4 min) | 3.1830 (59.3 min) * |
| 44 | 3.2034 (58.5 min) | 3.1818 (59.3 min) |
| 45 | 3.2026 (58.7 min) | 3.1824 (59.0 min) |

\* the submitted run. Gain on the same seed: 0.019 / 0.018 / 0.022 / 0.020; fixed-seed run-to-run
noise is about 0.001. The seed-45 baseline is this script with `--ttt-every 0 --dropout 0.1
--stoch-depth 0.05`, which restores the baseline loop. Published record: 3.195 at 59.0 min; our
reproduction of the baseline on this hardware is 0.005-0.008 worse than the
published number.

The episode is a faster optimizer for most of training rather than a better end point: the per-pass
validation loss is 0.09 below the baseline after the first pass, 0.02-0.04 through the middle, and
meets it in the last two passes; the three-checkpoint average then gains more than the baseline's.
Only the every-step form does this (every 16, 8, or 4 steps gain 0.002-0.003, every 2 steps 0.005).
Two controls at the same step schedule: a random direction of the same norm gives the unchanged-script
loss (3.1998), the ascent direction is worse than it (3.2035); consecutive-batch pairing (adapt on
this batch, test on the next) gives 3.1948 in twice the time. Step norms 0.3-0.5 work and 1.0 hurts;
the split ratio of the two halves does not matter as long as the optimizer receives the plain
full-batch mean.

Timing: the same script's training time depends on the node. On the nodes we used, identical
runs span 58.8-59.8 minutes; one node is 3 percent slower per step and runs there exceed the cap by
about half a minute at the same loss, so only runs from normal-speed nodes are listed. Peak memory
51.7 GB per GPU (baseline: 51.0 GB).

Script: `train.py` on current `main` with the episode added (+67 / -10 lines). Defaults reproduce
this entry except the seed; `--ttt-every 0 --dropout 0.1 --stoch-depth 0.05` gives back the baseline
training loop (the upstream micro-batch loop is kept verbatim in the `else` branch). Reproduce:

```bash
torchrun --standalone --nproc_per_node=8 train.py --seed 43
# equivalent explicit flags: --ttt-every 1 --ttt-step-norm 0.5 --ttt-step-schedule lr --ttt-plastic matrix_all --dropout 0.05 --stoch-depth 0.0
```

Training logs of every run in the table: https://gist.github.com/dangxingyu/c1c6754312714772b2927b8c6ca0c874 (one `Result saved to <path>` line per log redacted).

### CODE REVIEW

**Comment by akshayvegesna** [comment]
Left review on #99 which applies to this also, we can merge after those updates are made. 

Thanks, this is very cool!

**Comment by dangxingyu** [comment]
Updated per the review on #99: README row removed and the `ttt` flags renamed to `mg` (`--mg-every`, `--mg-step-norm`, ...). The tuning summary and the answer on the normalized step are in #99.
